cairo/programs/fork.cairo

```diff
@@ -0,0 +1,203 @@
+        let gas_used_delta = parent_gas_used - parent_gas_target;
+        let parent_fee_gas_delta = parent_base_fee_per_gas * gas_used_delta;
+        let (target_fee_gas_delta, _) = unsigned_div_rem(parent_fee_gas_delta, parent_gas_target);
+        let (base_fee_per_gas_delta, _) = unsigned_div_rem(
+            target_fee_gas_delta, BASE_FEE_MAX_CHANGE_DENOMINATOR
+        );
+        if (base_fee_per_gas_delta == 0) {
+            return 1;
+        }
+        return base_fee_per_gas_delta;
+    }
+
+    let gas_used_delta = parent_gas_target - parent_gas_used;
+    let parent_fee_gas_delta = parent_base_fee_per_gas * gas_used_delta;
+    let (target_fee_gas_delta, _) = unsigned_div_rem(parent_fee_gas_delta, parent_gas_target);
+    let (base_fee_per_gas_delta, _) = unsigned_div_rem(
+        target_fee_gas_delta, BASE_FEE_MAX_CHANGE_DENOMINATOR
+    );
+
+    return parent_base_fee_per_gas - base_fee_per_gas_delta;
+}
+
+// @notice See https://github.com/ethereum/execution-specs/blob/master/src/ethereum/cancun/fork.py#L288-L332
+// @dev Initial range checks for all values because header is filled with a hint
+func validate_header{range_check_ptr}(header: model.BlockHeader, parent_header: model.BlockHeader) {
+    // parent_hash
+    assert [range_check_ptr] = header.parent_hash.low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = header.parent_hash.high;
+    let range_check_ptr = range_check_ptr + 1;
+    // ommers_hash
+    assert [range_check_ptr] = header.ommers_hash.low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = header.ommers_hash.high;
+    let range_check_ptr = range_check_ptr + 1;
+    // coinbase
+    let (coinbase_high, coinbase_low) = split_felt(header.coinbase);
+    assert [range_check_ptr] = coinbase_low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = coinbase_high;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = 2 ** 32 - coinbase_high - 1;
+    let range_check_ptr = range_check_ptr + 1;
+    // state_root
+    assert [range_check_ptr] = header.state_root.low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = header.state_root.high;
+    let range_check_ptr = range_check_ptr + 1;
+    // transactions_root
+    assert [range_check_ptr] = header.transactions_root.low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = header.transactions_root.high;
+    let range_check_ptr = range_check_ptr + 1;
+    // receipt_root
+    assert [range_check_ptr] = header.receipt_root.low;
+    let range_check_ptr = range_check_ptr + 1;
+    assert [range_check_ptr] = header.receipt_root.high;
+    let range_check_ptr = range_check_ptr + 1;
+    // withdrawals_root
+    assert header.withdrawals_root.is_some * (1 - header.withdrawals_root.is_some) = 0;
+    let withdrawals_root = cast(header.withdrawals_root.value, Uint256*);
```

Comment:
Why a Uint256* ?
The withdrawals_root is an Option and Option is 
```
struct Option {
        is_some: felt,
        value: felt,
    }
```
https://github.com/kkrt-labs/keth/blob/e8e1da9291a2508d36d7f340d9df9339d8bba389/cairo/src/model.cairo#L222

Reply:
I might miss something as the same pattern is done for `parent_beacon_block_root` but not `base_fee_per_gas`

Reply:
`Option` is `Option<type>`, in this case it's a pointer to a `Uint256` (because hashes are uint256)

Reply:
you can check the types in https://github.com/alloy-rs/alloy/blob/main/crates/consensus/src/header.rs, some are `Option<u64>` so `felt` is enough, and some are `Option<B256>` so we need to have a `Uint256`, and so a `Uint256*`